writer.go

```diff
@@ -2,20 +2,189 @@ package warc
+// rawFile parameter.
+//
+// Seek will only be called with whence == io.SeekCurrent and offset == 0.
+//
+// See also CountWriter() if you need a "fake" Seek implementation.
+func NewWriterCompressed(rawFile io.WriteSeeker, cmprsWriter *gzip.Writer) (*Writer, error) {
+	w := &Writer{
+		seekW: rawFile,
+		wr:    cmprsWriter,
+	}
+	return w, nil
+}
+
+// NewWriterRaw initializes a WARC Writer writing to an uncompressed stream.
+// If the provided Writer implements io.Seeker, the RecordCallback will be
+// available.  If the provided Writer implements interface{Flush() error}, it
+// will be flushed after every written Record.
+//
+// See also CountWriter() if you need a "fake" Seek implementation.
+func NewWriterRaw(out io.Writer) (*Writer, error) {
+	w := &Writer{
+		wr: out,
+	}
+	if wseeker, ok := out.(io.WriteSeeker); ok {
+		w.seekW = wseeker
+	}
+	return w, nil
+}
+
+type countWriter struct {
+	count int64
+	w     io.Writer
+}
+
+// CountWriter implements a limited version of io.Seeker around the provided
+// Writer.  It only supports offset == 0 and whence == io.SeekCurrent or
+// io.SeekEnd, and returns the current number of written bytes in both cases.
+func CountWriter(w io.Writer) io.WriteSeeker {
+	return &countWriter{count: 0, w: w}
+}
+
+// implements io.Writer
+func (c *countWriter) Write(p []byte) (int, error) {
+	n, err := c.w.Write(p)
+	if n >= 0 {
+		c.count += int64(n)
+	}
+	return n, err
+}
+
+var errCountWriterNotImplemented = stdErrors.New("unsupported seek operation")
+
+// implements io.Seeker
+func (c *countWriter) Seek(offset int64, whence int) (int64, error) {
+	if offset != 0 || !(whence == io.SeekCurrent || whence == io.SeekEnd) {
+		return 0, errCountWriterNotImplemented
+	}
+	return c.count, nil
+}
+
+func (w *Writer) WriteRecord(r *Record) error {
+	var startPos, endPos int64
+	var err2 error
+	var errs []error
+
+	if w.seekW != nil {
+		startPos, err2 = w.seekW.Seek(0, io.SeekCurrent)
+		if err2 != nil {
+			errs = append(errs, errors.Wrap(err2, "warc writer: seek 0"))
+		}
+	}
+
+	// don't return quite yet - still need to do RecordCallback
+	err2 = r.Write(w.wr)
+	if err2 != nil {
+		errs = append(errs, errors.Wrap(err2, "warc writer: write record"))
+	}
+
+	if gzW, ok := w.wr.(*gzip.Writer); ok {
+		// flush is not sufficient, need to Close/Reset
+		err2 = gzW.Close()
+		if err2 != nil {
+			errs = append(errs, errors.Wrap(err2, "warc writer: flush"))
+		}
+		gzW.Reset(w.seekW)
+	}
+	if w.seekW != nil {
+		endPos, err2 = w.seekW.Seek(0, io.SeekCurrent)
+		if err2 != nil {
+			errs = append(errs, errors.Wrap(err2, "warc writer: seek 0"))
+		}
+	}
+
+	if w.RecordCallback != nil {
+		w.RecordCallback(r, startPos, endPos)
+	}
+
+	if len(errs) == 1 {
+		return errs[0]
+	} else if len(errs) > 1 {
+		return simpleMultiError(errs)
+	}
+	return nil
+}
+
+// Close cleans up any resources the warc.Writer might be holding on to.
+func (w *Writer) Close() error {
+	return nil
+}
+
+type simpleMultiError []error
```

Comment:
mind moving this into a separate file `errors.go`? Totally agreed that we need better error handling, pulling it into a separate file will help show when we need to break out a separate errors package.

Reply:
Making the start/end positions the function return value as mentioned above would completely remove the need for this error type - the whole reason was to avoid early returns so the callback still happened.